## Machine Learning Record Mining

Project to create a pipeline that uses GeoDeepDive's output to find Unaquired Sites for Neotoma.

Using NLP parsed text and a Data Science approach, identify whether a paper is suitable for Neotoma and detect features such as 'Site Name', 'Location', 'Age Span' and 'Site Descriptions'.

In [1]:
# Loading libraries

import numpy as np
import pandas as pd
import csv
import psycopg2
import re

# Load Postgres Server
from src.config import config

In [2]:
# Options for DF display
pd.set_option('display.max_colwidth', 50)
pd.set_option('display.max_rows', 10)

#pd.set_option('display.max_rows', None)
#pd.set_option('display.max_columns', None)
#pd.set_option('display.width', None)
#pd.set_option('display.max_colwidth', -1)

## Loading and viewing the Data

### Loading NLP Sentences

In [3]:
# Connect to PostgreSQL server from terminal:
# pg_ctl -D PSQL_Data -l logfile start

try:
    params = config()
    # Connect to the PostgreSQL database
    conn = psycopg2.connect(**params)
    # Create a new cursor
    cur = conn.cursor()
    nlp_sentences = pd.read_sql_query('''SELECT * FROM sentences;''', conn)
    # Close the cursor and connection to so the server can allocate
    # bandwidth to other requests
    cur.close()
    conn.close()
    print('Data obtained from PostgreSQL')
    
# If no SQL db, load from a file
except:
    header_list = ["_gddid", "sentid", "wordidx", "words", "part_of_speech", "special_class", 
               "lemmas", "word_type", "word_modified"]
    nlp_sentences = pd.read_csv("../Do_not_commit_data/sentences_nlp352", sep='\t', names = header_list)
    nlp_sentences = nlp_sentences.replace('"', '', regex = True)\
                                 .replace('\{', '', regex = True)\
                                 .replace('}', '', regex = True)\
                                 .replace(',', ',', regex = True)
    nlp_sentences['wordidx']= nlp_sentences['wordidx'].str.split(",")
    nlp_sentences['words']= nlp_sentences['words'].str.split(",")
    nlp_sentences['poses']= nlp_sentences['poses'].str.split(",")
    nlp_sentences['ners']= nlp_sentences['ners'].str.split(",")
    nlp_sentences['lemmas']= nlp_sentences['lemmas'].str.split(",")
    nlp_sentences['dep_paths']= nlp_sentences['dep_paths'].str.split(",")
    nlp_sentences['dep_parents']= nlp_sentences['dep_parents'].str.split(",")
    print('Data obtained from text file')

Data obtained from PostgreSQL


In [4]:
nlp_sentences.head(3)

_gddid  sentid  \
0  54b43266e138239d8684efed       1   
1  54b43266e138239d8684efed       2   
2  54b43266e138239d8684efed       3   

                                             wordidx  \
0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   
1            [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]   
2  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   

                                               words  \
0  [Available, online, at, www.sciencedirect.com,...   
1  [The, Chihuahueños, Bog, record, extends, to, ...   
2  [An, Artemisia, steppe, ,, then, an, open, Pic...   

                                      part_of_speech  \
0  [JJ, NN, IN, NNP, NNP, NNP, CD, -LRB-, CD, -RR...   
1  [DT, NNP, NN, NN, VBZ, TO, IN, CD, JJ, NN, NN, .]   
2  [DT, NNP, NN, ,, RB, DT, JJ, NNP, NN, VBD, IN,...   

                                       special_class  \
0  [O, O, O, O, O, O, NUMBER, O, DATE, O, NUMBER,...   
1   [O, O, O, O, O, O, O, NUMBER, O, DURATION, O, O]   
2  [O, O, O, O, O, O, O, LOCATION, LOCATION, O, O...   

                                              lemmas  \
0  [available, online, at, www.sciencedirect.com,...   
1  [the, Chihuahueños, bog, record, extend, to, o...   
2  [a, Artemisia, steppe, ,, then, a, open, Picea...   

                                           word_type  \
0  [dep, dep, dep, dep, dep, dep, dep, , dep, , d...   
1  [det, compound, compound, nsubj, , case, amod,...   
2  [det, compound, nsubj, , advmod, det, amod, co...   

                                       word_modified  
0  [218, 218, 218, 218, 218, 218, 218, 0, 218, 0,...  
1          [4, 4, 4, 5, 0, 11, 11, 11, 11, 11, 5, 0]  
2  [3, 3, 10, 0, 9, 9, 9, 9, 10, 0, 14, 14, 14, 1...

In [5]:
# Running this takes forever. Do not run unless necessary
# nlp_sentences.describe(include = 'all')

In [6]:
avg_sentences = nlp_sentences.groupby('_gddid')['_gddid','sentid'].transform('max')
avg_sentences.drop_duplicates(subset='_gddid', keep='first', inplace=True)
avg_sentences.describe()

sentid
count   218.000000
mean    513.600917
std     232.460891
min       2.000000
25%     397.250000
50%     504.000000
75%     638.250000
max    1330.000000

From the Summary Statistics, despite having 111,965 sentences, there are only 218 unique articles (`_gddid`). The article with most sentences, had 1330 sentences; that's the max number found in `sentid`.  
In order to know the average of sentences, I grouped by `_gddid` and then, obtained the number of sentences per article. After that, I just got the mean through a Summary of statistics to know that on average, each article has 514 sentences.

### Loading Bibliography Data

In [7]:
import json
with open('../Do_not_commit_data/bibjson', 'r') as f:
    bib_dict = json.load(f)

In [8]:
# Normalizing data so that we have access to the 'identifier'

# TODO Load into SQL server and connect through SQL
bibliography = pd.io.json.json_normalize(bib_dict,'identifier',['publisher', 'title', ['journal','name', 'name'], ['author'], 'year', 'number', 'volumne', ['link'], '_gddid', 'type', 'pages'], record_prefix='_', errors='ignore')

bibliography['link'] = bibliography['link'].astype(str)

url = bibliography['link'].str.split(", ", expand = True)

bibliography['link_url'] = url[0]
bibliography['link_type'] = url[1]

bibliography['link_url'] = bibliography['link_url'].replace("\[{'url': '", "", regex = True)\
                                                   .replace("'", "", regex = True)

bibliography['link_type'] = bibliography['link_type'].replace("'type': '", "", regex = True)\
                                                     .replace("'}]", "", regex = True)

bibliography['author'] = bibliography['author'].astype(str)
bibliography['author'] = bibliography['author'].replace("\[{'name': '", "", regex = True)\
                                               .replace("{'name': '", "", regex = True)\
                                               .replace("'},", ";", regex = True)\
                                               .replace("'}]", "", regex = True)
bibliography = bibliography[['_type', '_id', 'publisher', 'title', 'journal.name.name',	'author', 'year', 'number', 'volumne', '_gddid', 'type', 'pages', 'link_url', 'link_type']]

bibliography.head(3)

_type                            _id                    publisher  \
0   doi  10.1016/S0277-3791(99)00007-4                     Elsevier   
1   doi                10.1139/b92-002  Canadian Science Publishing   
2   doi   10.1016/0031-0182(80)90043-7                     Elsevier   

                                               title  \
0  Palaeoclimate, chronology and vegetation histo...   
1  Holocene history of forest trees on the Bruce ...   
2  Glacial sequence and environmental history in ...   

                                   journal.name.name  \
0                         Quaternary Science Reviews   
1                         Canadian Journal of Botany   
2  Palaeogeography, Palaeoclimatology, Palaeoecology   

                                              author  year number volumne  \
0  Huntley, B.; Watts, W.A.; Allen, J.R.M.; Zolit...  1999      7     NaN   
1                                     Bennett, K. D.  1992      1     NaN   
2  Van Der Hammen, T.; Barelds, J.; De Jong, H.; ...  1980    NaN     NaN   

                     _gddid     type     pages  \
0  550453fde1382326932d85f7  article  945--960   
1  578b5aabcf58f1587003efba  article     6--18   
2  54b43244e138239d868493cd  article  247--340   

                                            link_url  link_type  
0  http://www.sciencedirect.com/science/article/p...  publisher  
1  http://www.nrcresearchpress.com/doi/abs/10.113...  publisher  
2  http://www.sciencedirect.com/science/article/p...  publisher

In [9]:
bibliography.describe(include = 'all')

_type                               _id publisher  \
count    221                               221       221   
unique     1                               221         6   
top      doi  10.1111/j.1502-3885.2010.00159.x  Elsevier   
freq     221                                 1       138   

                                                    title  \
count                                                 221   
unique                                                221   
top     Charcoal signatures defined by multivariate an...   
freq                                                    1   

          journal.name.name          author  year number volumne  \
count                   221             221   221    183       0   
unique                   41             213    47     28       0   
top     Quaternary Research  Markgraf, Vera  2011      3     NaN   
freq                     44               3    14     40     NaN   

                          _gddid     type  pages  \
count                        221      221    216   
unique                       221        1    214   
top     550505d7e1382326932d8b07  article  1--13   
freq                           1      221      2   

                                                 link_url  link_type  
count                                                 221        220  
unique                                                221          1  
top     http://www.sciencedirect.com/science/article/p...  publisher  
freq                                                    1        220

## EDA

Reviewing our data includes skimming through some papers online and seeing if the data is consistent with our NLP Sentences dataframe.

From there, we can also visualize what we would like our model to predict: 'Location', 'Site Name', 'Age Span', and 'Site Description' from a "Human perspective".

### Finding Latitude and Longitud via REGEX 

After reviewing some PDF's and viewing the data parsing, I noticed some possible patterns for coordinates.

<img src="figures/img/00_snippet_of_pdf.png" alt="Drawing" style="width: 300px;"/>

Which translates to our pandas dataframe as:

![alt text](figures/img/01_snippet_in_sql.png "PandasVersion")

As a first problem, we can notice that the degrees symbol `°` was translated as a `3`. On some other occassions, it is confused as an `o` or a middle point `◦`. For `'`, this symbol got confused by the word `cents`.

I am still going to use normal REGEX to find coordinates in the most intuitive way. 

A coordinate for longitude is a number ranging from -180° to 180°. It has three components, `°, ',"` and it covers the position in E or W.

Latitude goes from -90° to 90°, has the same three components and goes from S to N.

Using REGEX, and assuming data consistency, the following code should be able to extract coordinates. 

In [10]:
dms_regex = r"([-]?\d{1,3}[°|′|\'|,]{0,3}\d{1,2}[,|′|\']{1,3}\d{0,2}[,|′|\']{1,4}[NESWnesw]?[\s|,|\']+?[-]?\d{1,3}[°|,|′|\']+\d{1,2}[,|′|\']+\d{0,2}[,|′|\'][,|′|\']{0,4}[NESWnesw]?)"

dd_regex = "[\\{,][-]?[1]?[0-9]{1,2}\\.[0-9]{1,}[,]?[NESWnesw],"

In [27]:
# TODO Find REGEX of couples of numbers (123, 234)
nlp_sentences['words_as_string'] = nlp_sentences['words'].apply(lambda x: ','.join(map(str, x)))
nlp_sentences['dms_regex'] = nlp_sentences['words_as_string'].str.findall(dms_regex)


#nlp_sentences[['dms_regex', 'words']]


#nlp_sentences['dms_regex'] = nlp_sentences['dms_regex'].astype(str)
#nlp_sentences['dms_regex'] = nlp_sentences['dms_regex'].replace(",", ";", regex = True)#\
                                                       #.replace(",", "", regex = True)

nlp_sentences[['words_as_string', 'words', 'dms_regex']].head(3)

words_as_string  \
0  Available,online,at,www.sciencedirect.com,Quat...   
1  The,Chihuahueños,Bog,record,extends,to,over,15...   
2  An,Artemisia,steppe,,,then,an,open,Picea,woodl...   

                                               words dms_regex  
0  [Available, online, at, www.sciencedirect.com,...        []  
1  [The, Chihuahueños, Bog, record, extends, to, ...        []  
2  [An, Artemisia, steppe, ,, then, an, open, Pic...        []

In [28]:
#nlp_sentences[['words_as_string', 'words', 'dms_regex']].describe()

The `dms_regex` only returned 95 unique values. That's to say that from the 111965 sentences, only 95 possess something similar to the regex we are looking for.  

Let's see an example of a regex straction as a sanity check that we are getting the values that we need. 

In [29]:
italy_example = nlp_sentences[nlp_sentences['_gddid'] == '550453fde1382326932d85f7']
italy_example = italy_example[italy_example['sentid'] == 10]
italy_example[['words', 'words_as_string', 'dms_regex']]

words  \
94875  [Introduction, Laghi, di, Monticchio, are, two...   

                                         words_as_string  \
94875  Introduction,Laghi,di,Monticchio,are,two,maar,...   

                         dms_regex  
94875  [403,56,40,N,,,153,36,48,E]

### Problems with the REGEX approach

As seen before, extracting information using REGEX is very complicated. We would require a lot of queries that would retrieve each particular case.

A possible alternative could be using NER, Named Entity Recognition.

In [30]:
def order_article(article_id):
    '''
    Function to find an article by its gddid in the NLP sentences and have it displayed in order
    
    Keyword arguments:
    article_id -- gddid
    
    Returns:
    article ordered by sentence index
    '''
    article = nlp_sentences[nlp_sentences['_gddid'] == article_id]
    return article[['sentid', 'words']].sort_values(by = 'sentid')

Article 01  
Skimmed info: 
 * `Article Name:` Palaeoclimate_chronology_and_vegetation_history_of_the_Weichselian_Lateglacial
 * `Link`: http://www.sciencedirect.com/science/article/pii/S0277379199000074  
 * `Site Name`:  Laghi di Monticchio
 * `Location`:  40° 56' 40" N, 15° 36' 48" E
 * `Age Span`: 
 * `Site Descriptions`: The record from Lago Grande di Monticchio now adds macrofossil evidence of the presence of relatively boreal tree taxa during the Lateglacial interstadial and provides a chronology based upon annually laminated sediments.

In [31]:
# Article 01 
order_article('550453fde1382326932d85f7')

sentid                                              words
94865       1  [Quaternary, Science, Reviews, 18, -LRB-, 1999...
94866       2  [Plant, macrofossil, data, from, one, core, pr...
94867       3  [Quantitative, palaeoclimate, reconstructions,...
94868       4  [The, chronological, framework, for, Lateglaci...
94870       5  [Surface, conditions, in, the, North, Atlantic...
...       ...                                                ...
95325     448  [Vegetation, history, and, climate, of, the, l...
95326     449  [Quaternary, Science, Reviews, ,, 15, ,, 133, ...
95327     450  [Zolitschka, ,, B., ,, &, Negendank, ,, J., F....
95329     451  [Sedimentology, ,, dating, and, palaeoclimatic...
95330     452  [Quaternary, Science, Reviews, ,, 15, ,, 101, ...

[452 rows x 2 columns]

Article 02  
Skimmed info:  
* `Article:` Development of the Mixed Connifer Forest in Northern New Mexico and its relationship to Holocene environmental change
* `Link`: http://www.sciencedirect.com/science/article/pii/S0033589407001512
* `Site Name`:  Chihuahuenos Bog 
* `Location`:  ??? No Given Coordinates 
* `Age Span`:   
* `Site Descriptions`:  Chihuahueños Bog is located within the mixed conifer forest and has affinities to the flora of the southern Rocky Mountains(Foxx and Tierney, 1985; Allen, 2004). Although the area immediately surrounding the bog has been heavily logged in recent years, nearly every conifer species growing in montaneforests of the Jemez mountains is represented either along theshore of the bog or in the surrounding uplands.  

In [32]:
# Article 2
order_article('54b43266e138239d8684efed')

sentid                                              words
0         1  [Available, online, at, www.sciencedirect.com,...
1         2  [The, Chihuahueños, Bog, record, extends, to, ...
2         3  [An, Artemisia, steppe, ,, then, an, open, Pic...
3         4  [C/N, ratios, ,, δ13C, and, δ15N, values, indi...
4         5  [Higher, percentages, of, aquatic, algae, and,...
..      ...                                                ...
605     572  [Is, the, Valles, caldera, entering, a, new, c...
144     573                  [Geology, 23, ,, 411, --, 414, .]
607     574  [Wright, Jr., ,, H.E., ,, Bent, ,, A.M., ,, Ha...
608     575  [Present, and, past, vegetation, of, the, Chus...
609     576  [Geological, Society, of, America, Bulletin, 8...

[576 rows x 2 columns]

Article 03  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/g35541.1  # No Full access to article
* `Site Name`:  Cahokia region, Illinois, USA
* `Location`:  N38°42'00" - N38°42'00", W90°04'60" - W90°04'60" - Not found on text (yet).
* `Age Span`:   
* `Site Descriptions`: Record from Horseshoe Lake, an oxbow lake in the central Mississippi River valley that is adjacent to the Cahokia site (Illinois, USA), the largest prehistoric settlement north of Mexico.   

In [33]:
# Article 3
order_article('57c5b941cf58f1338eaddb5b')

sentid                                              words
67708       1  [A, record, of, sustained, prehistoric, and, h...
67709       2  [Here, we, report, a, high-resolution, and, mu...
67710       3  [Palynological, and, carbon, isotope, data, do...
67712       4  [Rapid, forest, clearance, was, followed, clos...
67713       5  [Agricultural, intensification, that, included,...
...       ...                                                ...
67845     123  [Simon, ,, M.L., ,, and, Parker, ,, K.E., ,, 2...
67846     124  [Smith, ,, B.D., ,, and, Yarnell, ,, R.A., ,, ...
67848     125  [Sugita, ,, S., ,, 1993, ,, A, model, of, poll...
67849     126  [Trubitt, ,, M.B.D., ,, 2000, ,, Mound, buildi...
67850     127  [Woods, ,, W.I., ,, 2004, ,, Population, nucle...

[127 rows x 2 columns]

Article 04  
Skimmed info:  
* `Link:`http://www.tandfonline.com/doi/abs/10.1080/00173130902965157
* `Site Name`:  Begbunar
* `Location`:  42°09′ N, 22° 33′ E; 1750 ; in SQL file shows as: `42,°,09,cents,N,,,22,°,33,cents,E`
* `Age Span`:   
* `Site Descriptions`: Located in the central treeless zone of the Osogovo mountains, which are situated at the border between south-western Bulgaria and the north-eastern former Yugoslav Republic of Macedonia.

In [34]:
# Article 4
order_article('58d29193cf58f14928755ba5')

sentid                                              words
110659       1  [Grana, ISSN, :, 0017-3134, -LRB-, Print, -RRB...
110660       2  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...
110661       3  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...
110662       4  [Submit, your, article, to, this, journal, Art...
110665       5  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...
...        ...                                                ...
110740      80                                           [Ann, .]
110741      81                                  [Sofia, Univ., .]
110742      82                                           [Fac, .]
110743      83                                          [Geol, .]
110744      84  [Geogr., ,, 85, ,, 181, --, 198, -LRB-, in, Bu...

[84 rows x 2 columns]

In [35]:
# Try to see the exact location. Wont show last few words.
pd.DataFrame(order_article('58d29193cf58f14928755ba5').iloc[4])

110665
sentid                                                  5
words   [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...

Article 05  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co
* `Site Name`:  Cottonwood III, Cottonwood IV, Red Lady, Red Well, Splains
* `Location`:  38°49'50' 106°24'45', 38°49'50' 106°24'45', 38°52'50' 107°2'30', 38°52'50' 107°2'30',38°53'40' 107°3'15', 38°52'28' 107°05', 38°50' 107°4'30', 38°50' 107°4'30', 38°52' 107°2'30', 38°45' 106°50', 38°45' 106°50'
* `Age Span`:   17000 - 400
* `Site Descriptions`:   The lower timberline or lower forest border is the lower elevational limit of a continuous forest belt. In the Rocky Mountains it forms a sharp contrast with shrublands or grasslands in valley bottoms. Trees may extend below the lower timberline in favorable habitats, along streams or on rocky outcrops.  

In [36]:
# Article 5
order_article('57928e07cf58f133d1c26609')

sentid                                              words
39606       1  [Timberline, fluctuations, and, late, Quaterna...
39607       2  [By, tracking, climatically, sensitive, forest...
39608       3  [Pollen, data, suggest, that, prior, to, 11, 0...
39609       4  [The, inferred, climate, was, 2Ð5, ¡, C, coole...
39611       5  [Abies, -LRB-, fir, -RRB-, increased, in, abun...
...       ...                                                ...
40107     448  [Weber, ,, W., A., ,, 1987, ,, Colorado, flora...
40108     449  [Whitlock, ,, C., ,, 1993, ,, Postglacial, veg...
40110     450  [Whitlock, ,, C., ,, and, Bartlein, ,, P., J.,...
40111     451  [Wright, ,, H., E., ,, Jr., ,, 1983, ,, Late-Q...
40112     452  [MANUSCRIPT, RECEIVED, BY, THE, SOCIETY, MAY, ...

[452 rows x 2 columns]

In [37]:
# Print the exact line where locations are.
pd.DataFrame(order_article('57928e07cf58f133d1c26609').iloc[179])

39801
sentid                                                180
words   [LOCATION, AND, AGE, OF, FOSSIL, STUDY, SITES,...

In [38]:
regex_example = nlp_sentences[nlp_sentences['_gddid'] == '57928e07cf58f133d1c26609']
regex_example = regex_example[regex_example['sentid'] == 180]
regex_example[['words', 'words_as_string', 'dms_regex']]

words  \
39801  [LOCATION, AND, AGE, OF, FOSSIL, STUDY, SITES,...   

                                         words_as_string  \
39801  LOCATION,AND,AGE,OF,FOSSIL,STUDY,SITES,Core,Lo...   

                                               dms_regex  
39801  [38,°,49,′,50,′,′,106,°,24,′,45,′,′,, 38,°,49,...

In [39]:
# Find the article in the bibliography by gddid to retrieve Article Name

bibliography[bibliography['_gddid'] == '57928e07cf58f133d1c26609']

_type                                              _id publisher  \
93   doi  10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co;2       GSA   

                                                title  \
93  Timberline fluctuations and late Quaternary pa...   

                         journal.name.name             author  year number  \
93  Geological Society of America Bulletin  Fall, Patricia L.  1997     10   

   volumne                    _gddid     type       pages  \
93     NaN  57928e07cf58f133d1c26609  article  1306--1320   

                                             link_url  link_type  
93  http://dx.doi.org/10.1130/0016-7606(1997)109<1...  publisher

## Linking Neotoma DB to extract locations

In [40]:
neotoma = pd.read_csv("../Do_not_commit_data/data-1590729612420.csv")
neotoma = neotoma[['siteid', 'sitename', 'longitudeeast', 'latitudenorth', 'longitudewest', 'latitudesouth', 'sitedescription', 'doi']]
#neotoma.drop_duplicates(subset ="sitename", keep = False, inplace = True) 
neotoma

## Add strings for longitudeast and longitudenorth
## Then group by siteid, sitename and coordinates

siteid                  sitename  longitudeeast  latitudenorth  \
0      10330             Lac du Sommet      -70.66468       47.71662   
1      10330             Lac du Sommet      -70.66468       47.71662   
2       1729               Myrtle Lake      -93.37853       47.98645   
3       1729               Myrtle Lake      -93.37853       47.98645   
4        269              Billy's Lake      -94.54948       46.27186   
...      ...                       ...            ...            ...   
6884   26896  King & Kapp 1963 Site 19      -80.43829       47.49101   
6885   26896  King & Kapp 1963 Site 19      -80.43829       47.49101   
6886     525               Cupola Pond      -91.08914       36.79962   
6887   26913   Belaya Gora Horned Lark      147.16055       68.57887   
6888   26913   Belaya Gora Horned Lark      147.16055       68.57887   

      longitudewest  latitudesouth  \
0         -70.66573       47.71382   
1         -70.66573       47.71382   
2         -93.39207       47.97876   
3         -93.39207       47.97876   
4         -94.55363       46.26930   
...             ...            ...   
6884      -80.43829       47.49101   
6885      -80.43829       47.49101   
6886      -91.09082       36.79864   
6887      147.16055       68.57887   
6888      147.16055       68.57887   

                                        sitedescription  \
0     The small shallow Lac du Sommet (0.02 km2, 4 m...   
1     The small shallow Lac du Sommet (0.02 km2, 4 m...   
2     Lake surrounded by peatland. Physiography: Red...   
3     Lake surrounded by peatland. Physiography: Red...   
4     Shallow depression on St. Croix moraine. Physi...   
...                                                 ...   
6884  Site is near Makobe Lake, just north of the so...   
6885  Site is near Makobe Lake, just north of the so...   
6886  Small sinkhole pond almost completely occupied...   
6887  The site comprises a series of tunnels that ha...   
6888  The site comprises a series of tunnels that ha...   

                            doi  
0      10.1177/0959683611400199  
1      10.1177/0959683611400199  
2               10.1139/b68-190  
3               10.1139/b68-190  
4               10.2307/1939818  
...                         ...  
6884            10.1139/b63-022  
6885            10.1139/b63-022  
6886            10.2307/2845468  
6887  10.1038/s42003-020-0806-7  
6888  10.1038/s42003-020-0806-7  

[6889 rows x 8 columns]

In [41]:
nlp_bib = nlp_sentences.merge(bibliography, on = '_gddid')

In [42]:
nlp_bib[['words_as_string', 'journal.name.name', 'dms_regex']].head(10)

words_as_string    journal.name.name  \
0  Available,online,at,www.sciencedirect.com,Quat...  Quaternary Research   
1  The,Chihuahueños,Bog,record,extends,to,over,15...  Quaternary Research   
2  An,Artemisia,steppe,,,then,an,open,Picea,woodl...  Quaternary Research   
3  C/N,ratios,,,δ13C,and,δ15N,values,indicate,bot...  Quaternary Research   
4  Higher,percentages,of,aquatic,algae,and,elevat...  Quaternary Research   
5                           Anderson,,,R.S.,,,1989,.  Quaternary Research   
6  From,ca.,8500,to,6400,cal,yr,BP,the,pond,desic...  Quaternary Research   
7  C/N,ratios,declined,to,their,lowest,Holocene,l...  Quaternary Research   
8  Wetter,conditions,returned,after,6400,cal,yr,B...  Quaternary Research   
9  Higher,charcoal,influx,rates,after,6400,cal,yr...  Quaternary Research   

  dms_regex  
0        []  
1        []  
2        []  
3        []  
4        []  
5        []  
6        []  
7        []  
8        []  
9        []

In [43]:
print(nlp_bib.shape)

(111965, 24)


In [44]:
# Filter df to just frames that has REGEX longer than one
mask = (nlp_bib['dms_regex'].str.len() > 0)
# print(mask.shape)
nlp_bib = nlp_bib.loc[mask]
nlp_bib[['words_as_string','dms_regex']]

words_as_string  \
3314    Lake,Hope,Simpson,-LRB-,52,',27,',N,,,56,',26,...   
3318    The,lake,basin,itself,has,an,irregular,though,...   
3320    Moraine,Lake,-LRB-,52,',16,',N,,,58,',03,',W,,...   
3324    St.,Paul,Lake,-LRB-,52,',05,',N,,,57,',50,',W,...   
3408    Diagram,of,pollen,percentages,-LRB-,above,-RRB...   
...                                                   ...   
107854  Within,the,basin,,,members,of,the,DeForest,For...   
109253  Bob,Black,Pond,and,Quicksand,Pond,in,Bartow,Co...   
109436  C,III,--,I,,,no.,5,,,78,p.,MANUSCRIPT,RECEIVED...   
109480  Nothofagus,antarctica,,,Pilgerodendron,uviferu...   
110284  doi,:10.1016,/,j.yqres,.2006.08.006,298,Z.,Yu,...   

                                                dms_regex  
3314                          [52,',27,',N,,,56,',26,',W]  
3318                          [52,',33,',N,,,56,',54,',W]  
3320                          [52,',16,',N,,,58,',03,',W]  
3324                          [52,',05,',N,,,57,',50,',W]  
3408                          [52,',16,',N,,,58,',03,',W]  
...                                                   ...  
107854  [410,45,',N,410,40,',, 90,0,45,',90,0,40,',, 9...  
109253                                     [2,2,,,9,0,0,]  
109436                                      [28,,,1,9,7,]  
109480                             [710,26,',W,410,44,'S]  
110284                        [41,°,00,′,N,,,74,°,55,′,W]  

[99 rows x 2 columns]

Let's now try to extract Sitenames using Neotoma database. For this, I will group by doi and then, list all sitenames that are mentioned in each doi.

In [51]:
def list_items(df, col):
    return list(df[col])

In [53]:
#list_items(neotoma, 'sitename')

In [90]:
by_doi = neotoma.groupby('doi')
neotoma_sitenames = by_doi.apply(lambda x: list(x['sitename']))
neotoma_siteid = by_doi.apply(lambda x: list(x['siteid']))
neotoma_longeast = by_doi.apply(lambda x: list(x['longitudeeast']))
neotoma_latnorth = by_doi.apply(lambda x: list(x['latitudenorth']))

neotoma_joined_df = neotoma_sitenames.to_frame('sitenames').join(neotoma_siteid.to_frame('siteid')).join(neotoma_longeast.to_frame('longeast')).join(neotoma_latnorth.to_frame('latnorth'))
neotoma_joined_df = neotoma_joined_df.reset_index()

neotoma_joined_df['intlongeast'] = neotoma_joined_df['longeast'].astype(int)
neotoma_joined_df.head(5)


ValueError: setting an array element with a sequence.

From Neotoma DB, we have obtained a dataframe grouped by DOI that lists each `sitename`, `siteid`, `longitude east`, and `latitunde north`. Now, let's join this information to the NLP_Bib dataframe. For now, let's do to the one that coordinates have been obtained with REGEX just to see if some of the numbers and sitenames make sense. 